cmd/createorupdate/createorupdate.go

```diff
@@ -93,16 +94,38 @@ func delete(ctx context.Context, log *logrus.Entry, rpc sdk.OpenShiftManagedClus
 }
 
 func createOrUpdate(ctx context.Context, log *logrus.Entry, rpc sdk.OpenShiftManagedClustersClient, resourceGroup, manifest string) error {
+	aadClientID := os.Getenv("AZURE_AAD_CLIENT_ID")
+	clientID := os.Getenv("AZURE_CLIENT_ID")
+	aadClientSecret := os.Getenv("AZURE_CLIENT_SECRET")
+	if len(aadClientID) > 0 && aadClientID != clientID {
+		log.Info("updating the aad application")
+		if len(os.Getenv("AZURE_USERNAME")) == 0 || len(os.Getenv("AZURE_PASSWORD")) == 0 {
```

Comment:
I would suggest moving all aad-related stuff into its own function and call it in main after `if cleanup` and before the first `createorupdate`. The challenge there is to determine whether this is a create or update. You can achieve that either by trying to read `containerservice.yaml` or by doing a dummy check against the resourcegroup (if not exists, it's a create). The latter is probably a more realistic check since the client should not have access to `containerservice.yaml`.

Reply:
aad seems to authenticate differently. this works today as we use "az" which is a logged in user.
I struggled with getting this working and got some help from @jim-minter to get the initial login working.